File: 2022/go/02.go
package main

import (
	"fmt"
	"strings"
)

type Day2 struct {
}

type roundValues struct {
	opponent string
	me       string
}

type handShape int
type strategy int

const (
	drawHand     handShape = 0
	rockHand     handShape = 1
	paperHand    handShape = 2
	scissorsHand handShape = 3
)

const (
	winStrategy  strategy = 0
	drawStrategy strategy = 1
	loseStrategy strategy = 2
)

const (
	loseScore = 0
	drawScore = 3
	winScore  = 6
)

func (*Day2) Day() int {
	return 2
}

func (*Day2) IsTest() bool {
	return false
}

func (puzzle *Day2) Run(input string) {
	roundValues := puzzle.createRoundValuesList(input)

	puzzle.part1(roundValues)
	puzzle.part2(roundValues)
}

func (*Day2) getWinner(opponent, me handShape) handShape {
	if opponent == me {
		return drawHand
	}

	if opponent == rockHand && me == paperHand ||
		opponent == paperHand && me == scissorsHand ||
		opponent == scissorsHand && me == rockHand {
		return me
	}

	return opponent
}

func (*Day2) getWinHandShape(opponent handShape) handShape {
	if opponent == rockHand {
		return paperHand
	} else if opponent == paperHand {
		return scissorsHand
	}
	return rockHand
}

func (*Day2) getLoseHandShape(opponent handShape) handShape {
	if opponent == rockHand {
		return scissorsHand
	} else if opponent == paperHand {
		return rockHand
	}
	return paperHand
}

func (*Day2) convertHandShape(character string) handShape {
	if character == "A" || character == "X" {
		return rockHand
	} else if character == "B" || character == "Y" {
		return paperHand
	} else if character == "C" || character == "Z" {
		return scissorsHand
	}

	panic(fmt.Sprintf("Invalid round character %q", character))
}

func (*Day2) convertStrategy(character string) strategy {
	if character == "X" {
		return loseStrategy
	} else if character == "Y" {
		return drawStrategy
	} else if character == "Z" {
		return winStrategy
	}

	panic(fmt.Sprintf("Invalid round character %q", character))
}

func (puzzle *Day2) getHandShapeFromStrategy(opponent handShape, strategy strategy) handShape {
	if strategy == winStrategy {
		return puzzle.getWinHandShape(opponent)
	} else if strategy == drawStrategy {
		return opponent
	} else if strategy == loseStrategy {
		return puzzle.getLoseHandShape(opponent)
	}

	panic(fmt.Sprintf("Unknown strategy %v", strategy))
}

func (puzzle *Day2) getRoundScore(opponent, me handShape) int {
	winner := puzzle.getWinner(opponent, me)

	if opponent == me {
		return drawScore
	} else if winner == me {
		return winScore
	} else {
		return loseScore
	}
}

func (puzzle *Day2) part1(roundValues []roundValues) {
	score := 0

	for _, roundValue := range roundValues {
		opponent := puzzle.convertHandShape(roundValue.opponent)
		me := puzzle.convertHandShape(roundValue.me)

		score += puzzle.getRoundScore(opponent, me) + (int)(me)
	}

	fmt.Printf("Part1: score is %d\n", score)
}

func (puzzle *Day2) part2(roundValues []roundValues) {
	score := 0

	for _, roundValue := range roundValues {
		opponent := puzzle.convertHandShape(roundValue.opponent)
		strategy := puzzle.convertStrategy(roundValue.me)
		me := puzzle.getHandShapeFromStrategy(opponent, strategy)

		score += puzzle.getRoundScore(opponent, me) + (int)(me)
	}

	fmt.Printf("Part2: score is %d\n", score)
}

func (*Day2) createRoundValuesList(input string) []roundValues {
	result := []roundValues{}

	for _, line := range strings.Split(input, "\n") {
		parts := strings.Split(strings.TrimSpace(line), " ")
		result = append(result, roundValues{opponent: parts[0], me: parts[1]})
	}

	return result
}

File: 2022/cs/02.cs
namespace Day2Namespace;

class Day02 : IPuzzle
{
    private const int RoundLostScore = 0;
    private const int RoundDrawScore = 3;
    private const int RoundWinScore = 6;

    private record struct HandShapeInfo(int ShapeValue);

    private enum HandShape
    {
        Draw,
        Rock,
        Paper,
        Scissors,
    }

    private enum Strategy
    {
        Lost,
        Draw,
        Win,
    }

    private static HandShape GetHandShape(char letter)
    {
        switch (letter)
        {
            case 'A':
            case 'X':
                return HandShape.Rock;
            case 'B':
            case 'Y':
                return HandShape.Paper;
            case 'C':
            case 'Z':
                return HandShape.Scissors;
        }

        throw new InvalidDataException($"Unknown character '{letter}' in input data.");
    }

    private static Strategy GetStrategy(char letter)
    {
        switch (letter)
        {
            case 'X':
                return Strategy.Lost;
            case 'Y':
                return Strategy.Draw;
            case 'Z':
                return Strategy.Win;
        }

        throw new InvalidDataException($"Invalid strategy character '{letter}' in input data.");
    }

    private static HandShape GetWinner(HandShape opponent, HandShape me)
    {
        // int delta = (int)opponent - (int)me;

        // if (delta == 0)
        //     return HandShape.Draw;

        // if (delta == 1 || delta < -1)
        //     return opponent;

        // return me;

        if (opponent == me)
            return HandShape.Draw;

        if (opponent == HandShape.Rock)
            return me == HandShape.Paper ? me : opponent;

        if (opponent == HandShape.Paper)
            return me == HandShape.Scissors ? me : opponent;

        return me == HandShape.Rock ? me : opponent;
    }

    private static int GetRoundScore(HandShape opponent, HandShape me)
    {
        HandShape winner = GetWinner(opponent, me);

        if (winner == HandShape.Draw)
            return RoundDrawScore;

        if (winner == me)
            return RoundWinScore;

        return RoundLostScore;
    }

    private static HandShape GetWiningHandShape(HandShape opponent)
    {
        if (opponent == HandShape.Rock)
            return HandShape.Paper;
        if (opponent == HandShape.Paper)
            return HandShape.Scissors;
        if (opponent == HandShape.Scissors)
            return HandShape.Rock;

        throw new ArgumentException(nameof(opponent));
    }

    private static HandShape GetLoosingHandShape(HandShape opponent)
    {
        if (opponent == HandShape.Rock)
            return HandShape.Scissors;
        if (opponent == HandShape.Paper)
            return HandShape.Rock;
        if (opponent == HandShape.Scissors)
            return HandShape.Paper;

        throw new ArgumentException(nameof(opponent));
    }

    private static void Part1(List<(char opponentChar, char myChar)> entries)
    {
        int totalScore = 0;

        foreach ((char opponentChar, char myChar) in entries)
        {
            HandShape opponent = GetHandShape(opponentChar);
            HandShape me = GetHandShape(myChar);

            int roundScore = GetRoundScore(opponent, me);
            roundScore += (int)me;

            totalScore += roundScore;
        }

        Console.WriteLine($"My total score in part 1 is {totalScore}.");
    }

    private static HandShape GetPart2HandShape(HandShape opponent, Strategy strategy)
    {
        switch (strategy)
        {
            case Strategy.Lost:
                return GetLoosingHandShape(opponent);
            case Strategy.Draw:
                return opponent;
            case Strategy.Win:
                return GetWiningHandShape(opponent);
        }

        throw new ArgumentException(nameof(strategy));
    }

    private static void Part2(List<(char opponentChar, char myChar)> entries)
    {
        int totalScore = 0;

        foreach ((char opponentChar, char myChar) in entries)
        {
            HandShape opponent = GetHandShape(opponentChar);
            HandShape me = GetPart2HandShape(GetHandShape(opponentChar), GetStrategy(myChar));

            int roundScore = GetRoundScore(opponent, me);
            roundScore += (int)me;

            totalScore += roundScore;
        }

        Console.WriteLine($"My total score in part 2 is {totalScore}.");
    }

    public int Day => 2;
    public bool IsTest => false;

    public void Run(string input)
    {
        int lineNumber = -1;
        List<(char, char)> entries = new();

        foreach (string line in input.Split('\n').Select(x => x.Trim()))
        {
            lineNumber++;

            if (string.IsNullOrWhiteSpace(line))
                continue;

            string[] parts = line.Split(' ');

            if (parts.Length != 2 || parts[0].Length != 1 || parts[1].Length != 1)
                throw new InvalidDataException($"Invalid data at line '{lineNumber}'.");

            entries.Add((parts[0][0], parts[1][0]));
        }

        Part1(entries);
        Part2(entries);
    }
}
